core/src/main/scala/filodb.core/metadata/Projection.scala

```diff
@@ -24,28 +27,59 @@ case class Projection(id: Int,
+ * It is also guaranteed to have a valid sortColumn and dataset partition column, and will
+ * have a SortKeyHelper and partitioning function as well.
  */
-case class RichProjection(id: Int,
-                          dataset: Dataset,
-                          sortColumn: Column,
-                          reverse: Boolean,
-                          columns: Seq[Column]) {
-  def helper[K]: SortKeyHelper[K] = Dataset.sortKeyHelper(dataset, sortColumn).get
+case class RichProjection[K](id: Int,
+                             dataset: Dataset,
+                             sortColumn: Column,
+                             sortColNo: Int,
+                             reverse: Boolean,
+                             columns: Seq[Column],
+                             helper: SortKeyHelper[K],
+                             partitionFunc: RowReader => PartitionKey) {
+  def sortKeyFunc: RowReader => K = helper.getSortKeyFunc(Seq(sortColNo))
 }
 
 object RichProjection {
-  // TODO: error handling
-  def apply(dataset: Dataset, columns: Seq[Column], projectionId: Int = 0): RichProjection = {
+  case class BadSchema(reason: String) extends Exception("BadSchema: " + reason)
+
+  def apply[K](dataset: Dataset, columns: Seq[Column], projectionId: Int = 0): RichProjection[K] = {
+    val tryProj = make[K](dataset, columns, projectionId)
+    tryProj.recover {
+      case t: Throwable => throw t
+    }
+    tryProj.get
+  }
+
+  def make[K](dataset: Dataset, columns: Seq[Column], projectionId: Int = 0): Try[RichProjection[K]] = {
+    def fail(reason: String): Try[RichProjection[K]] = Failure(BadSchema(reason))
+
     val normProjection = dataset.projections(projectionId)
     val richColumns = {
       if (normProjection.columns.isEmpty) {
         columns
       } else {
         val columnMap = columns.map { c => c.name -> c }.toMap
+        val missing = normProjection.columns.toSet -- columnMap.keySet
+        if (missing.nonEmpty) return fail(s"Specified projection columns are missing: $missing")
         normProjection.columns.map(columnMap)
       }
     }
-    val sortColumn = richColumns.filter(_.hasId(normProjection.sortColumn)).head
-    RichProjection(projectionId, dataset, sortColumn, normProjection.reverse, richColumns)
+
+    val sortColNo = richColumns.indexWhere(_.hasId(normProjection.sortColumn))
+    if (sortColNo < 0) return fail(s"Sort column ${normProjection.sortColumn} not in columns $richColumns")
+
+    val sortColumn = richColumns(sortColNo)
+    if (!(SortKeyHelper.ValidSortClasses contains sortColumn.columnType.clazz)) {
+      return fail(s"Unsupported sort column type ${sortColumn.columnType}")
+    }
+
+    val helper = Dataset.sortKeyHelper[K](dataset, sortColumn).get
+
+    for { partitionFunc <- Dataset.getPartitioningFunc(dataset, richColumns) } yield {
+      RichProjection[K](projectionId, dataset, sortColumn, sortColNo,
+                        normProjection.reverse, richColumns, helper, partitionFunc)
+    }
   }
```

Comment:
I'd break this out to smaller private functions and remove the need to call 'return'.